File: golang/examples/schema-registry/producer.go
package main

import (
	"encoding/binary"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
	"github.com/linkedin/goavro/v2"
	"github.com/riferrei/srclient"
)

// serializeAvroWithSchemaID сериализует сообщение с использованием Avro и добавляет Schema ID
func serializeAvroWithSchemaID(codec *goavro.Codec, schemaID int, message map[string]interface{}) ([]byte, error) {
	// Сериализуем сообщение с использованием Avro
	binaryData, err := codec.BinaryFromNative(nil, message)
	if err != nil {
		return nil, fmt.Errorf("ошибка при сериализации сообщения: %w", err)
	}

	// Добавляем Magic Byte и Schema ID в начало сообщения
	schemaIDBytes := make([]byte, 4)
	binary.BigEndian.PutUint32(schemaIDBytes, uint32(schemaID))

	// Формат сообщения: Magic Byte (1 байт) + Schema ID (4 байта) + Avro данные
	payload := append([]byte{0}, schemaIDBytes...)
	payload = append(payload, binaryData...)

	return payload, nil
}

// setupProducer создает и настраивает Kafka продюсера
func setupProducer(logger *log.Logger) (*kafka.Producer, error) {
	// Конфигурация продюсера
	config := &kafka.ConfigMap{
		"bootstrap.servers": "kafka:29092",
	}

	// Создаем продюсера
	producer, err := kafka.NewProducer(config)
	if err != nil {
		return nil, fmt.Errorf("ошибка при создании продюсера: %w", err)
	}

	// Обработка отчетов о доставке в отдельной горутине
	go func() {
		for e := range producer.Events() {
			switch ev := e.(type) {
			case *kafka.Message:
				if ev.TopicPartition.Error != nil {
					logger.Printf("Ошибка доставки: %v", ev.TopicPartition.Error)
				} else {
					logger.Printf("Сообщение доставлено в %s [%d] со смещением %v",
						*ev.TopicPartition.Topic, ev.TopicPartition.Partition, ev.TopicPartition.Offset)
				}
			}
		}
	}()

	return producer, nil
}

func main() {
	// Создаем логгер
	logger := log.New(os.Stdout, "avro-producer: ", log.LstdFlags)
	logger.Println("Запуск продюсера с Avro и Schema Registry...")

	// Название топика
	topic := "avro-test-topic"

	// Определяем Avro схему для наших сообщений
	avroSchemaJSON := `{
		"type": "record",
		"name": "Message",
		"namespace": "com.example",
		"fields": [
			{"name": "id", "type": "int"},
			{"name": "content", "type": "string"},
			{"name": "timestamp", "type": "long"},
			{"name": "title", "type": ["null", "string"], "default": null}
		]
	}`

	// Создаем клиент для Schema Registry
	schemaRegistryClient := srclient.CreateSchemaRegistryClient("http://schema-registry:8081")
	logger.Println("Подключение к Schema Registry...")

	// Регистрируем схему в Schema Registry
	schema, err := schemaRegistryClient.CreateSchema(topic+"-value", avroSchemaJSON, srclient.Avro)
	if err != nil {
		logger.Fatalf("Ошибка при регистрации схемы: %v", err)
	}
	logger.Printf("Схема зарегистрирована с ID: %d", schema.ID())

	// Создаем кодек для сериализации Avro
	codec, err := goavro.NewCodec(avroSchemaJSON)
	if err != nil {
		logger.Fatalf("Ошибка при создании Avro кодека: %v", err)
	}

	// Создаем и настраиваем продюсера
	producer, err := setupProducer(logger)
	if err != nil {
		logger.Fatalf("Ошибка при настройке продюсера: %v", err)
	}
	defer producer.Close()

	// Сообщения для отправки
	messages := []map[string]interface{}{
		{
			"id":        1,
			"content":   fmt.Sprintf("Первое сообщение с Avro: %s", time.Now().Format(time.RFC3339)),
			"timestamp": time.Now().Unix(),
			"title":     "Заголовок для первого сообщения",
		},
		{
			"id":        2,
			"content":   fmt.Sprintf("Второе сообщение с Avro: %s", time.Now().Format(time.RFC3339)),
			"timestamp": time.Now().Unix(),
			"title":     nil, // Явно указываем nil для демонстрации nullable поля
		},
		{
			"id":        3,
			"content":   fmt.Sprintf("Третье сообщение с Avro: %s", time.Now().Format(time.RFC3339)),
			"timestamp": time.Now().Unix(),
			"title":     "Еще один заголовок",
		},
	}

	// Отправляем сообщения
	for i, msg := range messages {
		logger.Printf("Подготовка сообщения: %v", msg)

		// Сериализуем сообщение с использованием Avro и добавляем Schema ID
		payload, err := serializeAvroWithSchemaID(codec, schema.ID(), msg)
		if err != nil {
			logger.Printf("Ошибка при сериализации сообщения: %v", err)
			continue
		}

		// Отправляем сообщение
		err = producer.Produce(&kafka.Message{
			TopicPartition: kafka.TopicPartition{
				Topic:     &topic,
				Partition: kafka.PartitionAny,
			},
			Value: payload,
			Key:   []byte(fmt.Sprintf("key-%d", i)),
		}, nil)

		if err != nil {
			logger.Printf("Ошибка при отправке сообщения: %v", err)
			continue
		}

		logger.Printf("Отправка сообщения #%d", i+1)

		// Даем время на обработку сообщения
		time.Sleep(500 * time.Millisecond)
	}

	// Ждем, пока все сообщения будут отправлены
	remaining := producer.Flush(15000)
	if remaining > 0 {
		logger.Printf("Предупреждение: %d сообщений не были отправлены", remaining)
	} else {
		logger.Println("Все сообщения успешно отправлены!")
	}
}

File: php/examples/schema-registry/producer.php
<?php

require_once __DIR__ . '/../../vendor/autoload.php';

use FlixTech\AvroSerializer\Objects\RecordSerializer;
use FlixTech\SchemaRegistryApi\Registry\CachedRegistry;
use FlixTech\SchemaRegistryApi\Registry\PromisingRegistry;
use FlixTech\SchemaRegistryApi\Registry\Cache\AvroObjectCacheAdapter;
use GuzzleHttp\Client;
use Monolog\Logger;
use Monolog\Handler\StreamHandler;
use AvroSchema;

// Создаем логгер
$logger = new Logger('kafka-avro-producer');
$logger->pushHandler(new StreamHandler('php://stdout', Logger::DEBUG));

// Название топика
$topic = 'avro-test-topic';

// Определяем Avro схему для наших сообщений
$avroSchemaJson = <<<SCHEMA
{
    "type": "record",
    "name": "Message",
    "namespace": "com.example",
    "fields": [
        {"name": "id", "type": "int"},
        {"name": "content", "type": "string"},
        {"name": "timestamp", "type": "long"},
        {"name": "title", "type": ["null", "string"], "default": null}
    ]
}
SCHEMA;

// Создаем объект AvroSchema из JSON строки
$avroSchema = AvroSchema::parse($avroSchemaJson);

try {
    // Создаем клиент для Schema Registry
    $schemaRegistryClient = new Client(['base_uri' => 'http://schema-registry:8081']);

    // Создаем реестр схем
    $registry = new CachedRegistry(
        new PromisingRegistry($schemaRegistryClient),
        new AvroObjectCacheAdapter()
    );

    // Создаем сериализатор для Avro с опцией автоматической регистрации схем
    $recordSerializer = new RecordSerializer($registry, [
        RecordSerializer::OPTION_REGISTER_MISSING_SUBJECTS => true,
        RecordSerializer::OPTION_REGISTER_MISSING_SCHEMAS => true
    ]);

    // Конфигурация продюсера
    $conf = new RdKafka\Conf();
    $conf->set('metadata.broker.list', 'kafka:29092');

    // Создаем продюсера
    $producer = new RdKafka\Producer($conf);

    // Получаем топик
    $kafkaTopic = $producer->newTopic($topic);

    // Сообщения для отправки
    $messages = [
        [
            'id' => 1,
            'content' => 'Первое сообщение с Avro: ' . date('Y-m-d H:i:s'),
            'timestamp' => time(),
            'title' => 'Заголовок для первого сообщения'
        ],
        [
            'id' => 2,
            'content' => 'Второе сообщение с Avro: ' . date('Y-m-d H:i:s'),
            'timestamp' => time(),
            'title' => null // Явно указываем null для демонстрации nullable поля
        ],
        [
            'id' => 3,
            'content' => 'Третье сообщение с Avro: ' . date('Y-m-d H:i:s'),
            'timestamp' => time(),
            'title' => 'Еще один заголовок'
        ]
    ];

    // Отправляем сообщения
    foreach ($messages as $index => $message) {
        $logger->info("Подготовка сообщения: " . json_encode($message));

        // Сериализуем сообщение с использованием Avro и Schema Registry
        $encodedMessage = $recordSerializer->encodeRecord(
            $topic . '-value', // Суффикс -value для схемы значений
            $avroSchema,
            $message
        );

        $logger->info("Отправка сообщения #{$message['id']}");

        // RD_KAFKA_PARTITION_UA означает, что брокер сам выберет раздел
        $kafkaTopic->produce(RD_KAFKA_PARTITION_UA, 0, $encodedMessage, "key-{$index}");

        // Даем время на обработку сообщения
        $producer->poll(0);
    }

    // Ждем, пока все сообщения будут отправлены
    for ($flushRetries = 0; $flushRetries < 10; $flushRetries++) {
        $result = $producer->flush(10000);
        if (RD_KAFKA_RESP_ERR_NO_ERROR === $result) {
            break;
        }
    }

    if (RD_KAFKA_RESP_ERR_NO_ERROR !== $result) {
        $logger->error("Не удалось отправить все сообщения. Код ошибки: {$result}");
    } else {
        $logger->info('Все сообщения успешно отправлены!');
    }
} catch (Exception $e) {
    $logger->error('Ошибка: ' . $e->getMessage());
    $logger->error($e->getTraceAsString());
}
